main.js

```diff
@@ -0,0 +1,271 @@
+let animationTimer = false;
+let res = 10
+let points = [];
+
+/* INITIAL STATE */
+let state = {
+  controlsOpen: false,
+  numPoints: 1200,
+  speed: 0.025,
+  showFlow: true,
+  paused: false,
+};
+
+// user controls
+const controls = {
+  numPoints: {
+    type: "text",
+    display: "number of lines",
+  },
+  showFlow: {
+    type: "checkbox",
+    display: "show flow fields"
+  }
+};
+
+// run the thang
+init()
+
+/* SET STATE FUNCTION */
+function setStateAndDraw(nextState) {
+  state = Object.assign({}, state, nextState);
+  console.log("state update", state);
+  draw();
+}
+
+function setStateOnly(nextState, callback = null) {
+  state = Object.assign({}, state, nextState);
+  console.log("state update", state);
+  callback;
+}
+
+function init() {
+  console.log("state:", state)
+
+  // d3.select("#container")
+  d3.select("#canvas")
+    .on("click", () => pauseOrPlay())
+
+  controlContainer = d3.select("div#controlsContainer")
+    .style("width", `${controlsWidth + controlPanelWidth}px`)
+    .style("right", `-${controlPanelWidth}px`)
+
+  const replay = controlContainer
+    .append("div")
+    .attr("class", "replay")
+    .style("right", `-${controlPanelWidth}px`)
+    .on("click", () => {
+      draw()
+    })
+    .append("img")
+    .attr("src", "./assets/replay.svg")
+    .style("width", `${controlsIconWidth}px`)
+    .style("height", `${controlsIconWidth}px`)
+    
+  const save = controlContainer
+    .append("div")
+    .attr("class", "save")
+    .style("right", `-${controlPanelWidth}px`)
+    .on("click", () => {
+      saveCanvasImage()
+    }).append("img")
+    .attr("src", "./assets/save.svg")
+    .style("width", `${controlsIconWidth}px`)
+    .style("height", `${controlsIconWidth}px`)
+
+  const cog = controlContainer
+    .append("div")
+    .attr("class", "cog")
+    .style("right", `-${controlPanelWidth}px`)
+    .on("click", () => {
+      setStateOnly({ controlsOpen: !state.controlsOpen }, toggleControlPanel())
+    }).append("img")
+    .attr("src", "./assets/cog.svg")
+    .style("width", `${controlsIconWidth}px`)
+    .style("height", `${controlsIconWidth}px`)
+
+  const controlPanel = controlContainer
+    .append("div")
+    .attr("class", "controlPanel")
+    .style("width", `${controlsWidth + controlPanelWidth}px`)
+    .style("right", `-${controlPanelWidth}px`)
+
+  const controlItems = controlPanel
+    .selectAll(".controlItem")
+    .data(Object.entries(controls))
+    .join("div")
+    .attr("class", "controlItem")
+
+  controlItems
+    .append("input")
+    .attr("type", ([_, details]) => details.type)
+    .attr("class", ([_, details]) => details.type)
+    .property("checked", ([name, _]) => state[name])
+    .on("click", ([name, details]) => {
+      details.type === 'text' 
+        ? null
+        : setStateAndDraw({ [name]: !state[name] })
+    })
+    // specific to the text input
+    .attr("size", 5)
+    .on("keypress", function([name, _]) {
+      d3.event.keyCode === 13 
```

Comment:
i prefer to use `if(){}` here, saving ternaries for assigning/evaluating a value